src/lsdb/loaders/hipscat/hipscat_loader_factory.py

```diff
@@ -0,0 +1,36 @@
+from typing import TypeVar
+
+from lsdb import Catalog
+from lsdb.catalog.dataset.dataset import Dataset
+from lsdb.loaders.hipscat.hipscat_catalog_loader import HipscatCatalogLoader
+from lsdb.loaders.hipscat.hipscat_loading_config import HipscatLoadingConfig
+
+
+CatalogTypeVar = TypeVar("CatalogTypeVar", bound=Dataset)
+
+
+loader_class_for_catalog_type = {
+    Catalog: HipscatCatalogLoader,
+}
+
+
+def get_loader_for_type(
+        catalog_type_to_use: CatalogTypeVar,
+        path: str,
+        config: HipscatLoadingConfig
+):
+    """Constructs a CatalogLoader that loads a Dataset of the specified type
+
+    Args:
+        catalog_type_to_use (Type[Dataset]): the type of catalog to be loaded. Uses the actual type
```

Comment:
Could we be generous, and accept those other types? Or is this a semi-protected method that will be called from within lsdb loaders?

Reply:
Again it's for type checking reasons. I don't think there's a way to get it to understand the type of catalog it returns depends on the value of a string, so if users wanted it to type correctly they'd have to use casts. I think this is better, even If it might be slightly unusual to users who don't do type checking.

Reply:
Having said that, there might be a way to overload the method so that if it uses the type it'll return the right type for people who care about type checking, or just returns the base type if it's a string. Let me look into that.

Reply:
Ok so I got lost in a bit of a type checking rabbit hole, but I've updated the code so that we use this less now. This method is unchanged, but this is just a semi-protected method and shouldn't be used by an end user.

The `read_hipscat` method though will be used by the user and had the same problem. I've now changed it so that by default it will read the catalog_info file and use the type specified in there. The problem with this is the method is typed to return the base `Dataset` class. It will still actually return the correct type of object, but the type checker won't pick that up (They can be different because python). The `catalog_type` parameter is only used if a user wants to have proper type checking and so they can specify the correct type to make sure that happens.

I kinda don't like this, but I think it's the best option. The alternative is having different read methods for different types of catalogs, like `read_association_catalog`, `read_object_catalog`, `read_source_catalog`. It would make type checking a lot easier, but it feels kinda not very pythonic and a bit clunky, although I it is better from an OOP perspective. I think we've discussed this before and wanted the single read method, but I'm not sure which is better.